File: code-gen-web-frontend/src/services/backend/generatorController.ts
// @ts-ignore
/* eslint-disable */
import { request } from '@umijs/max';

/** isFreeById GET /api/generator/${param0}/free */
export async function isFreeByIdUsingGet(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.isFreeByIdUsingGETParams,
  options?: { [key: string]: any },
) {
  const { id: param0, ...queryParams } = params;
  return request<API.BaseResponseBoolean_>(`/api/generator/${param0}/free`, {
    method: 'GET',
    params: { ...queryParams },
    ...(options || {}),
  });
}

/** addGenerator POST /api/generator/add */
export async function addGeneratorUsingPost(
  body: API.GeneratorAddRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseLong_>('/api/generator/add', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** deleteGenerator POST /api/generator/delete */
export async function deleteGeneratorUsingPost(
  body: API.DeleteRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseBoolean_>('/api/generator/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** editGenerator POST /api/generator/edit */
export async function editGeneratorUsingPost(
  body: API.GeneratorEditRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseBoolean_>('/api/generator/edit', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** getGeneratorVOById GET /api/generator/get/vo */
export async function getGeneratorVoByIdUsingGet(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getGeneratorVOByIdUsingGETParams,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseGeneratorVO_>('/api/generator/get/vo', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  });
}

/** listGeneratorByPage POST /api/generator/list/page */
export async function listGeneratorByPageUsingPost(
  body: API.GeneratorQueryRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponsePageGenerator_>('/api/generator/list/page', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** listGeneratorVOByPage POST /api/generator/list/page/vo */
export async function listGeneratorVoByPageUsingPost(
  body: API.GeneratorQueryRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponsePageGeneratorVO_>('/api/generator/list/page/vo', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** makeGeneratorOnline POST /api/generator/make */
export async function makeGeneratorOnlineUsingPost(
  body: API.MakingGeneratorRequest,
  options?: { [key: string]: any },
) {
  return request<any>('/api/generator/make', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** listMyGeneratorVOByPage POST /api/generator/my/list/page/vo */
export async function listMyGeneratorVoByPageUsingPost(
  body: API.GeneratorQueryRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponsePageGeneratorVO_>('/api/generator/my/list/page/vo', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** updateGenerator POST /api/generator/update */
export async function updateGeneratorUsingPost(
  body: API.GeneratorUpdateRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseBoolean_>('/api/generator/update', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** useGeneratorByIdOnline POST /api/generator/use */
export async function useGeneratorByIdOnlineUsingPost(
  body: API.UseGeneratorRequest,
  options?: { [key: string]: any },
) {
  return request<any>('/api/generator/use', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/** getGeneratorForUserPurchase GET /api/generator/users/purchase */
export async function getGeneratorForUserPurchaseUsingGet(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getGeneratorForUserPurchaseUsingGETParams,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseUserGenerator_>('/api/generator/users/purchase', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  });
}

File: code-gen-web-backend/src/main/java/com/pandaer/web/generator/controller/GeneratorController.java
package com.pandaer.web.generator.controller;

import cn.hutool.core.util.ObjectUtil;
import cn.hutool.core.util.ReUtil;
import cn.hutool.core.util.StrUtil;
import cn.hutool.json.JSONUtil;
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.pandaer.maker.meta.Meta;
import com.pandaer.maker.meta.MetaValidator;
import com.pandaer.web.common.annotation.AuthCheck;
import com.pandaer.web.common.dto.req.DeleteRequest;
import com.pandaer.web.common.dto.resp.BaseResponse;
import com.pandaer.web.common.enums.ErrorCode;
import com.pandaer.web.common.exception.BusinessException;
import com.pandaer.web.common.exception.ThrowUtils;
import com.pandaer.web.common.manager.CosManager;
import com.pandaer.web.common.utils.ResultUtils;
import com.pandaer.web.generator.dto.req.*;
import com.pandaer.web.generator.dto.resp.GeneratorFeeVO;
import com.pandaer.web.generator.dto.resp.GeneratorVO;
import com.pandaer.web.generator.entity.Generator;
import com.pandaer.web.generator.entity.GeneratorFee;
import com.pandaer.web.generator.entity.UserGenerator;
import com.pandaer.web.generator.enums.GeneratorStatus;
import com.pandaer.web.generator.service.GeneratorFeeService;
import com.pandaer.web.generator.service.GeneratorService;
import com.pandaer.web.generator.service.UserGeneratorService;
import com.pandaer.web.user.entity.User;
import com.pandaer.web.user.service.UserService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.BeanUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;

import javax.annotation.Resource;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import static com.pandaer.web.common.constant.UserConstant.ADMIN_ROLE;

/**
 * 帖子接口
 *
 
 */
@RestController
@RequestMapping("/generator")
@Slf4j
public class GeneratorController {

    @Resource
    private GeneratorService generatorService;


    @Resource
    private UserGeneratorService userGeneratorService;

    @Resource
    private UserService userService;

    @Resource
    private GeneratorFeeService generatorFeeService;

    @Autowired
    private CosManager cosManager;

    @Autowired
    private ApplicationEventPublisher applicationEventPublisher;

    // region 增删改查

    /**
     * 创建
     *
     * @param generatorAddRequest
     * @param request
     * @return
     */
    @PostMapping("/add")
    public BaseResponse<Long> addGenerator(@RequestBody GeneratorAddRequest generatorAddRequest, HttpServletRequest request) {
        if (generatorAddRequest == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }

        String name = generatorAddRequest.getName();
        // 额外的增强逻辑
        // 校验代码生成器名称是否符合Java命名规范
        if (!ReUtil.isMatch("^[a-zA-Z_$][a-zA-Z0-9_$]*$", name)) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR, "名称必须符合Java标识符规范（字母/下划线/$开头，后续可包含数字）");
        }

        // 补充校验：不能是Java关键字（可选增强）
        if (StrUtil.hasBlank(name) || ReUtil.isMatch("^(abstract|assert|boolean|break|byte|...)$", name)) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR, "名称不能是Java保留关键字");
        }

        Long generatorId = generatorService.addGenerator(generatorAddRequest, request);
        return ResultUtils.success(generatorId);
    }

    /**
     * 删除
     *
     * @param deleteRequest
     * @param request
     * @return
     */
    @PostMapping("/delete")
    public BaseResponse<Boolean> deleteGenerator(@RequestBody DeleteRequest deleteRequest, HttpServletRequest request) {
        if (deleteRequest == null || deleteRequest.getId() <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        User user = userService.getLoginUser(request);
        long id = deleteRequest.getId();
        // 判断是否存在
        Generator oldGenerator = generatorService.getById(id);
        ThrowUtils.throwIf(oldGenerator == null, ErrorCode.NOT_FOUND_ERROR);
        // 仅本人或管理员可删除
        if (!oldGenerator.getUserId().equals(user.getId()) && !userService.isAdmin(request)) {
            throw new BusinessException(ErrorCode.NO_AUTH_ERROR);
        }
        boolean b = generatorService.removeById(id);
        return ResultUtils.success(b);
    }

    /**
     * 更新（仅管理员）
     *
     * @param generatorUpdateRequest
     * @return
     */
    @PostMapping("/update")
    @AuthCheck(mustRole = ADMIN_ROLE)
    public BaseResponse<Boolean> updateGenerator(@RequestBody GeneratorUpdateRequest generatorUpdateRequest) {
        if (generatorUpdateRequest == null || generatorUpdateRequest.getId() <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        Generator generator = new Generator();
        BeanUtils.copyProperties(generatorUpdateRequest, generator);
        List<String> tags = generatorUpdateRequest.getTags();
        if (tags != null) {
            generator.setTags(JSONUtil.toJsonStr(tags));
        }
        // 参数校验
        generatorService.validGenerator(generator, false);
        long id = generatorUpdateRequest.getId();
        // 判断是否存在
        Generator oldGenerator = generatorService.getById(id);
        ThrowUtils.throwIf(oldGenerator == null, ErrorCode.NOT_FOUND_ERROR);
        boolean result = generatorService.updateById(generator);
        return ResultUtils.success(result);
    }

    /**
     * 在线使用代码生成器 TODO 需要抽离逻辑
     * @param useGeneratorRequest 在线使用代码生成器的参数
     * @return
     */
    @PostMapping("/use")
    public void useGeneratorByIdOnline(@RequestBody UseGeneratorRequest useGeneratorRequest,
                                       HttpServletRequest request,
                                       HttpServletResponse response) throws IOException, InterruptedException {

        // 用户登录
        User loginUser = userService.getLoginUser(request);
        if (loginUser == null) {
            throw new BusinessException(ErrorCode.NOT_LOGIN_ERROR);
        }

        // 获取用户参数
        Long generatorId = useGeneratorRequest.getId();
        Map<String, Object> dataModel = useGeneratorRequest.getDataModel();

        // 获取对应ID的代码生成器信息
        Generator generator = generatorService.getById(generatorId);
        if (generator == null) {
            throw new BusinessException(ErrorCode.NOT_FOUND_ERROR,"对应的代码生成器不存在");
        }

        String distPath = generator.getDistPath();
        if (StrUtil.isBlank(distPath)) {
            throw new BusinessException(ErrorCode.NOT_FOUND_ERROR,"对应的代码生成器产物包不存在");
        }

        // 使用代码生成器
        File generatedCodeFilesZip = generatorService.useGenerator(generator, distPath, dataModel);

        // 返回给前端
        response.setContentType(MediaType.APPLICATION_OCTET_STREAM_VALUE);
        String fileName = generatedCodeFilesZip.getName();
        response.setHeader("Content-Disposition", "attachment;filename=" +fileName );
        Files.copy(generatedCodeFilesZip.toPath(), response.getOutputStream());

    }



    /**
     * 根据 id 获取
     *
     * @param id
     * @return
     */
    @GetMapping("/get/vo")
    public BaseResponse<GeneratorVO> getGeneratorVOById(long id, HttpServletRequest request) {
        if (id <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        Generator generator = generatorService.getById(id);
        if (generator == null) {
            throw new BusinessException(ErrorCode.NOT_FOUND_ERROR);
        }
        return ResultUtils.success(generatorService.getGeneratorVO(generator, request));
    }

    /**
     * 分页获取列表（仅管理员）
     *
     * @param generatorQueryRequest
     * @return
     */
    @PostMapping("/list/page")
    @AuthCheck(mustRole = ADMIN_ROLE)
    public BaseResponse<Page<Generator>> listGeneratorByPage(@RequestBody GeneratorQueryRequest generatorQueryRequest) {
        long current = generatorQueryRequest.getCurrent();
        long size = generatorQueryRequest.getPageSize();

        // TODO 这样实现不够优雅
        QueryWrapper<Generator> queryWrapper = generatorService.getQueryWrapper(generatorQueryRequest);
//        queryWrapper.eq("status", GeneratorStatus.AUDIT_PASS.getValue());
        Page<Generator> generatorPage = generatorService.page(new Page<>(current, size), queryWrapper);
        return ResultUtils.success(generatorPage);
    }

    /**
     * 分页获取列表（封装类）
     *
     * @param generatorQueryRequest
     * @param request
     * @return
     */
    @PostMapping("/list/page/vo")
    public BaseResponse<Page<GeneratorVO>> listGeneratorVOByPage(@RequestBody GeneratorQueryRequest generatorQueryRequest, HttpServletRequest request) {
        long current = generatorQueryRequest.getCurrent();
        long size = generatorQueryRequest.getPageSize();
        // 限制爬虫
        ThrowUtils.throwIf(size > 20, ErrorCode.PARAMS_ERROR);


        // TODO 这样实现不够优雅
        QueryWrapper<Generator> queryWrapper = generatorService.getQueryWrapper(generatorQueryRequest);
        queryWrapper.eq("status", GeneratorStatus.AUDIT_PASS.getValue());
        Page<Generator> generatorPage = generatorService.page(new Page<>(current, size), queryWrapper);
        return ResultUtils.success(generatorService.getGeneratorVOPage(generatorPage, request));
    }

    /**
     * 分页获取当前用户创建的资源列表
     *
     * @param generatorQueryRequest
     * @param request
     * @return
     */
    @PostMapping("/my/list/page/vo")
    public BaseResponse<Page<GeneratorVO>> listMyGeneratorVOByPage(@RequestBody GeneratorQueryRequest generatorQueryRequest, HttpServletRequest request) {
        if (generatorQueryRequest == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        User loginUser = userService.getLoginUser(request);
        generatorQueryRequest.setUserId(loginUser.getId());
        long current = generatorQueryRequest.getCurrent();
        long size = generatorQueryRequest.getPageSize();
        // 限制爬虫
        ThrowUtils.throwIf(size > 20, ErrorCode.PARAMS_ERROR);
        Page<Generator> generatorPage = generatorService.page(new Page<>(current, size), generatorService.getQueryWrapper(generatorQueryRequest));
        return ResultUtils.success(generatorService.getGeneratorVOPage(generatorPage, request));
    }

    // endregion


    /**
     * 在线制作代码生成器
     * @param makingGeneratorRequest 制作代码生成器需要使用的参数
     * @param response 响应体
     */
    @PostMapping("/make")
    public void makeGeneratorOnline(@RequestBody MakingGeneratorRequest makingGeneratorRequest, HttpServletResponse response) {

        // 校验Meta元信息配置
        Meta meta = makingGeneratorRequest.getMeta();
        MetaValidator.validate(meta);

        String zipTemplateFilesUrl = makingGeneratorRequest.getZipTemplateFilesUrl();
        // TODO 校验 url是否存在

        // 一个校验的增强，校验代码生成器的名字只能是英文
        String name = makingGeneratorRequest.getMeta().getName();
        // 校验代码生成器名称是否符合Java命名规范
        if (!ReUtil.isMatch("^[a-zA-Z_$][a-zA-Z0-9_$]*$", name)) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR, "名称必须符合Java标识符规范（字母/下划线/$开头，后续可包含数字）");
        }

        // 补充校验：不能是Java关键字（可选增强）
        if (StrUtil.hasBlank(name) || ReUtil.isMatch("^(abstract|assert|boolean|break|byte|...)$", name)) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR, "名称不能是Java保留关键字");
        }

        generatorService.makeGenerator(makingGeneratorRequest,response);
    }


    @GetMapping("/users/purchase")
    public BaseResponse<UserGenerator> getGeneratorForUserPurchase(Long userId,Long generatorId) {
        if (ObjectUtil.hasNull(userId,generatorId)) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        UserGenerator purchasedGenerator = userGeneratorService.getGeneratorForPurchase(userId, generatorId);
        return ResultUtils.success(purchasedGenerator);
    }






    /**
     * 编辑（用户）
     *
     * @param generatorEditRequest
     * @param request
     * @return
     */
    @PostMapping("/edit")
    public BaseResponse<Boolean> editGenerator(@RequestBody GeneratorEditRequest generatorEditRequest, HttpServletRequest request) {
        if (generatorEditRequest == null || generatorEditRequest.getId() <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        Generator generator = generatorEditRequest.toGenerator();
        // 参数校验
        generatorService.validGenerator(generator, false);

        User loginUser = userService.getLoginUser(request);
        long id = generatorEditRequest.getId();
        // 判断是否存在
        Generator oldGenerator = generatorService.getById(id);
        ThrowUtils.throwIf(oldGenerator == null, ErrorCode.NOT_FOUND_ERROR);
        // 仅本人或管理员可编辑
        if (!oldGenerator.getUserId().equals(loginUser.getId()) && !Objects.equals(loginUser.getUserRole(), ADMIN_ROLE)) {
            throw new BusinessException(ErrorCode.NO_AUTH_ERROR);
        }
        boolean result = generatorService.updateById(generator);
        GeneratorFeeVO generatorFeeVO = generatorEditRequest.getGeneratorFee();
//        GeneratorFee generatorFee = BeanUtil.toBean(generatorFeeVO, GeneratorFee.class);
        if (generatorFeeVO!=null && generatorFeeVO.getPrice()!=null) {
            generatorFeeService.lambdaUpdate().eq(GeneratorFee::getGeneratorId,generator.getId()).set(GeneratorFee::getPrice,generatorFeeVO.getPrice()).update();
        }


        return ResultUtils.success(result);
    }


    @GetMapping("/{id}/free")
    public BaseResponse<Boolean> isFreeById(@PathVariable("id") Long generatorId) {
        if (generatorId == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }

        Generator generator = generatorService.getById(generatorId);
        if (generator == null) {
            throw new BusinessException(ErrorCode.NOT_FOUND_ERROR);
        }

        Boolean res = generatorService.isFreeById(generatorId);
        return ResultUtils.success(res);


    }

}
